### DISCUSSION TITLE
Backups/WAL archiving to local PVC using Barman server

### DISCUSSION THREAD

**Discussion by bo0tzz** [cloudnative-pg/cloudnative-pg#496] [Ideas]
The current barman-cloud backup system works well, but carries unnecessary complexity for small deployments where an object store is not already present (like in a homelab environment). It necessitates an extra deployment of MinIO ([also mentioned in the cnpg docs](https://cloudnative-pg.io/documentation/1.15.0/backup_recovery/#minio-gateway), bringing with it more things to manage and a potential extra point of failure. Additionally, the needed MinIO Gateway [is now deprecated](https://github.com/minio/minio/issues/14331).

It would be very nice to instead have the option of using a Barman server running inside the Kubernetes cluster. Ideally this server would be managed by the cnpg operator, although I can imagine the scope of that might be too large. Either way, it would simplify the backup setup and be nicer to use.

I think this could also have merit in larger deployments where someone is already using an (external) Barman server with postgres deployments that they want to migrate into cnpg - it would allow for bootstrapping from the existing Barman store, without first needing to shoehorn that data into an object store.

I imagine this has already been discussed during internal development before open-sourcing the project, but since I couldn't find a public discussion of the topic yet I thought I'd get one going.

Thanks for this wonderful tool!

**Comment by sxd** [comment]
Hi @bo0tzz 

I'm turning this into a discussion, since this could be the root of a couple of issues not just one

Regards!

**Comment by eshikhov** [comment]
Hi

A good idea.

This would also solve the problem with restoring without data loss RPO=0 ("zero data loss") .
I would like to be able to restore the cluster from a backup made by Barman with a non-cloud version or another tool. To be able to restore the database without losing data. Maybe someone already solved this problem and there is a workaround?

**Comment by gbartolini** [comment]
@bo0tzz yes, it has been discussed, internally at EDB before we open sourced and spawned CloudNativePG.

As you can understand, Barman is special to us (both EDB and our team, me in particular) and we'd definitely love to see a Barman operator land in the ecosystem as it would fill another gap in the use cases arena (like the one @eshikhov) is mentioning.

I can only talk about my team at EDB in terms of capacity: at the moment we do not have this in our near roadmap. However, if anyone else would like to try and develop this, we can definitely support with design and review of a Barman operator.

In terms of workaround, you can already integrate an external Barman system in the backup part, just relying on streaming replication (so, no rsync and no WAL archiving). In this case, the only missing bit is the possibility to specify in CloudNativePG additional synchronous names to be used in the list of sync standbys.

A different beast is however the recovery. The only existing workaround I can think of at the moment is to use Barman to restore the cluster, then connect from Kubernetes using the `pg_basebackup` recovery method. I know, it is terrible, but it will work.

In terms of recovery, what you could try doing (but needs to be tested) is:

- create a 1 instance cluster
- hibernate it
- use barman recover to replace the content of PGDATA
- exit hibernation
- cross your fingers

**Reply by vladimirfx** [reply]
What do you think about the 4th recovery method (let's name it `init_container` or `custom`)?
Configuration is just K8s init container spec. The outcome is the entire Pg data dir in the specified path. With such a method, any backup restore solution can be used.

**Reply by vladimirfx** [reply]
Maybe now possible to provide the Pg data dir for the operator-managed pod?
